lib/services/livesync/android-livesync-library.ts

```diff
@@ -0,0 +1,396 @@
+		this.handleConnection(connectionResult);
+	}
+
+	public async sendFile(filePath: string): Promise<void> {
+		await this.sendFileHeader(filePath);
+		await this.sendFileContent(filePath);
+	}
+
+	public async sendFiles(filePaths: string[]) {
+		for (const filePath of filePaths) {
+			if (this.$fs.getLsStats(filePath).isFile()) {
+				if (!this.$fs.exists(filePath)) {
+					this.$errors.fail(`${filePath} doesn't exist.`);
+				}
+
+				await this.sendFile(filePath);
+			}
+		}
+	}
+
+	public sendDirectory(directoryPath: string) {
+		const list = this.$fs.enumerateFilesInDirectorySync(directoryPath);
+		return this.sendFiles(list);
+	}
+
+	public removeFile(filePath: string): Promise<boolean> {
+		return new Promise((resolve: Function, reject: Function) => {
+			this.verifyActiveConnection(reject);
+			const filePathData = this.getFilePathData(filePath);
+			const headerBuffer = Buffer.alloc(PROTOCOL_OPERATION_LENGTH_SIZE +
+				SIZE_BYTE_LENGTH +
+				filePathData.filePathLengthSize +
+				filePathData.filePathLengthBytes);
+
+			let offset = 0;
+			offset += headerBuffer.write(DELETE_FILE_OPERATION.toString(), offset, PROTOCOL_OPERATION_LENGTH_SIZE);
+			offset = headerBuffer.writeInt8(filePathData.filePathLengthSize, offset);
+			offset += headerBuffer.write(filePathData.filePathLengthString, offset, filePathData.filePathLengthSize);
+			headerBuffer.write(filePathData.relativeFilePath, offset, filePathData.filePathLengthBytes);
+			const hash = crypto.createHash("md5").update(headerBuffer).digest();
+
+			this.socketConnection.write(headerBuffer);
+			this.socketConnection.write(hash, () => {
+				resolve(true);
+			});
+		});
+	}
+
+	public removeFiles(files: string[]) {
+		return Promise.all(files.map(file => this.removeFile(file)));
+	}
+
+	public generateOperationIdentifier(): string {
+		return crypto.randomBytes(16).toString("hex");
+	}
+
+	public isOperationInProgress(operationId: string): boolean {
+		return !!this.operationPromises[operationId];
+	}
+
+	public sendDoSyncOperation(operationId: string, timeout: number): Promise<any> {
+		const id = operationId || this.generateOperationIdentifier();
+		const operationPromise = new Promise((resolve: Function, reject: Function) => {
+			this.verifyActiveConnection(reject);
+
+			const message = `${DO_SYNC_OPERATION}${id}`;
+			const socketId = this.socketConnection.uid;
+			const hash = crypto.createHash("md5").update(message).digest();
+
+			this.operationPromises[id] = {
+				resolve,
+				reject,
+				socketId
+			};
+
+			this.socketConnection.write(message);
+			this.socketConnection.write(hash);
+
+			setTimeout(() => {
+				if (this.isOperationInProgress(id)) {
+					this.handleSocketError(socketId, "Sync operation is taking too long");
+				}
+			}, 60000);
+		});
+
+		return operationPromise;
+	}
+
+	public end() {
+		if (this.socketConnection) {
+			this.socketConnection.end();
```

Comment:
Can we also set `socketConnection` to null and allow a new `connect` call?

Reply:
This will happen in the close event callback.